db/change_cache.go

```diff
@@ -577,31 +577,118 @@ func (c *changeCache) releaseUnusedSequence(ctx context.Context, sequence uint64
+	if fromSequence == toSequence {
 		change := &LogEntry{
-			Sequence:     sequence,
+			Sequence:     toSequence,
 			TimeReceived: timeReceived,
 		}
-
-		// Since processEntry may unblock pending sequences, if there were any changed channels we need
-		// to notify any change listeners that are working changes feeds for these channels
 		changedChannels := c.processEntry(ctx, change)
 		allChangedChannels = allChangedChannels.Update(changedChannels)
 		c.channelCache.AddUnusedSequence(change)
+		if c.notifyChange != nil {
+			c.notifyChange(ctx, allChangedChannels)
+		}
+		return
 	}
 
+	// acquire mutex on cache for reading/writing to pending + reading next sequence value
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if toSequence < c.nextSequence {
+		// batch remove from skipped
+		err := c.RemoveSkippedSequences(ctx, fromSequence, toSequence)
+		if err != nil {
+			base.DebugfCtx(ctx, base.KeyCache, err.Error())
+			// we have sequences (possibly duplicate sequences) in the unused range that don't exist in skipped list
+			seqRange := &channels.UnusedSequenceRange{
+				StartSeq: fromSequence,
+				EndSeq:   toSequence,
+			}
+			// handle any potential duplicate sequences between the unused range and the skipped list
+			err = c.skippedSeqs._removeSubsetOfRangeFromSkipped(ctx, seqRange)
+			if err != nil {
+				base.DebugfCtx(ctx, base.KeyCache, "error processing subset of unused sequences in skipped list: %v", err)
+			}
+		}
+	} else if fromSequence >= c.nextSequence {
+		// whole range to pending
+		seqRange := &channels.UnusedSequenceRange{
+			StartSeq: fromSequence,
+			EndSeq:   toSequence,
+		}
+		c._pushRangeToPending(ctx, seqRange, timeReceived)
+		// unblock any pending sequences we can after new range(s) have been pushed to pending
+		changedChannels := c._addPendingLogs(ctx)
+		allChangedChannels = allChangedChannels.Update(changedChannels)
+		c.internalStats.pendingSeqLen = len(c.pendingLogs)
+	} else {
+		base.DebugfCtx(ctx, base.KeyCache, "unused sequence range of #%d to %d contains duplicate sequences", fromSequence, toSequence)
+	}
+	// update high seq cached
+	c.channelCache.AddUnusedSequence(&LogEntry{Sequence: toSequence})
+
 	if c.notifyChange != nil {
 		c.notifyChange(ctx, allChangedChannels)
 	}
 }
 
+// _pushRangeToPending will push a sequence range to pending logs. If pending has entries in it, we will check if
+// those entries are in the range and handle it, so we don't push duplicate sequences to pending
+func (c *changeCache) _pushRangeToPending(ctx context.Context, seqRange *channels.UnusedSequenceRange, timeReceived time.Time) {
+	if c.pendingLogs.Len() == 0 {
+		// push whole range & return early to avoid duplicate checks
+		entry := &LogEntry{
+			TimeReceived:  timeReceived,
+			Sequence:      seqRange.StartSeq,
+			SequenceRange: seqRange,
+		}
+		heap.Push(&c.pendingLogs, entry)
+		return
+	}
+
+	// check for duplicate sequences between range and pending logs
+	for _, v := range c.pendingLogs {
```

Comment:
c.pendingLogs can be large, and so this is an expensive way to check for intersections.  Since c.pendingLogs is sorted, can we instead use a more efficient search (e.g. sort.Find) to look for the first entry in pending that's greater than startSeq, and then check whether that value is less than endSeq?

Reply:
I have added this change to use sort.Find. Very good suggestion. I was trying to figure a way to do this same approach for skipped sequences in `_removeSubsetOfRangeFromSkipped` but I think this is difficult. But it may not be as much of a problem given we now have ranges in skipped list.

Reply:
I have now added a similar approach in `_removeSubsetOfRangeFromSkipped` using sort.Find to find minimum index in skipped list that has overlap in the unused sequence range, then iterate from that point on (avoiding iterating through more of the skipped list than we need to). Let me know what you think, it doesn't seem a clean solution to me but its what I could think of at this time.